src/tsconfig/validator.ts

```diff
@@ -0,0 +1,407 @@
+/* eslint-disable @typescript-eslint/member-ordering */
+
+/**
+ * A function that receives 3 arguments and validates if the provided value matches.
+ * @param value The value to validate
+ * @param reporter A function that will be called by the matcher with a a violation message.
+ * This function is always called, regardless of the outcome of the matcher.
+ * It is up to the caller of the matcher to decide if the message should be used or not.
+ * @param hints A function that might receive explicitly allowed values.
+ * This can be used to generate synthetics values that would match the matcher.
+ * It is not guaranteed that hints are received or that hints are complete.
+ *
+ * @returns true if the value matches
+ */
+type Matcher = (value: any, reporter: (message: string) => void, hints?: (allowed: any[]) => void) => boolean;
```

Comment:
💅 This could take an object instead of positional parameters, would be easier to extend in the future.